uc.c

```diff
@@ -935,206 +918,72 @@ MemoryRegion *memory_mapping(struct uc_struct* uc, uint64_t address)
-                                  *evh = i;
-                                  uc->hook_out_idx = i;
-                                  return UC_ERR_OK;
-                              } else
-                                  return UC_ERR_NOMEM;
-                     case UC_X86_INS_IN:
-                              // only one event handler at the same time
-                              if (uc->hook_in_idx)
-                                  return UC_ERR_HOOK_EXIST;
-
-                              i = hook_find_new(uc);
-                              if (i) {
-                                  uc->hook_callbacks[i].callback = callback;
-                                  uc->hook_callbacks[i].user_data = user_data;
-                                  *evh = i;
-                                  uc->hook_in_idx = i;
-                                  return UC_ERR_OK;
-                              } else
-                                  return UC_ERR_NOMEM;
-                     case UC_X86_INS_SYSCALL:
-                     case UC_X86_INS_SYSENTER:
-                              // only one event handler at the same time
-                              if (uc->hook_syscall_idx)
-                                  return UC_ERR_HOOK_EXIST;
-
-                              i = hook_find_new(uc);
-                              if (i) {
-                                  uc->hook_callbacks[i].callback = callback;
-                                  uc->hook_callbacks[i].user_data = user_data;
-                                  *evh = i;
-                                  uc->hook_syscall_idx = i;
-                                  return UC_ERR_OK;
-                              } else
-                                  return UC_ERR_NOMEM;
-                 }
-                 break;
-    }
-
-    return UC_ERR_OK;
-}
-
 UNICORN_EXPORT
 uc_err uc_hook_add(uc_engine *uc, uc_hook *hh, int type, void *callback, void *user_data, ...)
 {
     va_list valist;
     int ret = UC_ERR_OK;
-    int id;
-    uint64_t begin, end;
 
     va_start(valist, user_data);
 
-    if (type & UC_HOOK_MEM_READ_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_READ_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_WRITE_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_WRITE_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_FETCH_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_FETCH_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_READ_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_READ_PROT, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_WRITE_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_WRITE_PROT, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_FETCH_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_FETCH_PROT, callback, user_data, hh);
+    struct hook *hook = calloc(1, sizeof(struct hook));
```

Comment:
i dont see anywhere this "hook" is freed. list_remove() and list_clear() dont do this, so we are having memory leaking issue now?